src/librbd/cache/rwl/Request.cc

```diff
@@ -0,0 +1,361 @@
+  this->set_cell(guard_ctx.cell);
+}
+
+template <typename T>
+void C_WriteRequest<T>::finish_req(int r) {
+  ldout(rwl.get_context(), 15) << "write_req=" << this << " cell=" << this->get_cell() << dendl;
+
+  /* Completed to caller by here (in finish(), which calls this) */
+  utime_t now = ceph_clock_now();
+  rwl.release_write_lanes(this);
+  ceph_assert(m_resources.allocated);
+  m_resources.allocated = false;
+  this->release_cell(); /* TODO: Consider doing this in appending state */
+  update_req_stats(now);
+}
+
+template <typename T>
+void C_WriteRequest<T>::setup_buffer_resources(
+    uint64_t &bytes_cached, uint64_t &bytes_dirtied, uint64_t &bytes_allocated,
+    uint64_t &number_lanes, uint64_t &number_log_entries,
+    uint64_t &number_unpublished_reserves) {
+
+  ceph_assert(!m_resources.allocated);
+
+  auto image_extents_size = this->image_extents.size();
+  m_resources.buffers.reserve(image_extents_size);
+
+  bytes_cached = 0;
+  bytes_allocated = 0;
+  number_lanes = image_extents_size;
+  number_log_entries = image_extents_size;
+  number_unpublished_reserves = image_extents_size;
+
+  for (auto &extent : this->image_extents) {
+    m_resources.buffers.emplace_back();
+    struct WriteBufferAllocation &buffer = m_resources.buffers.back();
+    buffer.allocation_size = MIN_WRITE_ALLOC_SIZE;
+    buffer.allocated = false;
+    bytes_cached += extent.second;
+    if (extent.second > buffer.allocation_size) {
+      buffer.allocation_size = extent.second;
+    }
+    bytes_allocated += buffer.allocation_size;
+  }
+  bytes_dirtied = bytes_cached;
+}
+
+template <typename T>
+void C_WriteRequest<T>::setup_log_operations() {
+  {
+    std::lock_guard locker(m_lock);
+    // TODO: Add sync point if necessary
+    std::shared_ptr<SyncPoint> current_sync_point = rwl.get_current_sync_point();
+    uint64_t current_sync_gen = rwl.get_current_sync_gen();
+    op_set =
+      make_unique<WriteLogOperationSet>(this->m_dispatched_time,
+                                        m_perfcounter,
+                                        current_sync_point,
+                                        rwl.get_persist_on_flush(),
+                                        rwl.get_context(), this);
+    ldout(rwl.get_context(), 20) << "write_req=" << *this << " op_set=" << op_set.get() << dendl;
+    ceph_assert(m_resources.allocated);
+    /* op_set->operations initialized differently for plain write or write same */
+    for (auto &extent : this->image_extents) {
+      /* operation->on_write_persist connected to m_prior_log_entries_persisted Gather */
+      auto operation =
+        std::make_shared<WriteLogOperation>(*op_set, extent.first, extent.second, rwl.get_context());
+      op_set->operations.emplace_back(operation);
+    }
+    auto allocation = m_resources.buffers.begin();
+    uint64_t buffer_offset = 0;
+    for (auto &gen_op : op_set->operations) {
+      /* A WS is also a write */
+      auto operation = gen_op->get_write_op();
```

Comment:
Nit: drop `get_write_op` and replace w/ a generic handler method that can be properly overridden in the write op derived class. Alternatively, a few lines prior you created a `WriteLogOperationSet`, so just retain the derived pointer so that you have direct access to the necessary write ops w/o the need for casting nor magic helper methods.

Reply:
Done